### PR TITLE
Add configuration exception classes and exit codes

### PR DESCRIPTION
**Pull request by SpacePossum**
replacement of https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1500 on 1.12

### CODE REVIEW

**Comment by keradus** [comment]
What about other places? Like Header fixer?

**Comment by SpacePossum** [comment]
* HeaderCommentFixer can be done, todo
* PhpUnitConstructFixer done
* ExtraEmptyLinesFixer not possible, config options on master only

**Comment by keradus** [comment]
> ExtraEmptyLinesFixer not possible, config options on master only

Great to hear that!
In that case I ask you to apply that changes on 2.x line when this will be merged ;) Thanks!

**Comment by keradus** [comment]
Shouldn't the `InvalidFixerConfigurationException` be caught in `FixCommand` ?

**Comment by SpacePossum** [comment]
We could :) do you have something in mind other than the default exception display of the symfony/console package?

**Comment by keradus** [comment]
I would leave it to you to propose console output ;) Just remember to return correct exit status

**Comment by keradus** [comment]
ping :)

**Comment by keradus** [comment]
ping @SpacePossum :)

**Comment by keradus** [comment]
I would suggest to create namespace `Symfony\CS\ConfigurationException` and in it two classes:
- `InvalidConfigurationException`
- `InvalidFixerConfigurationException` extending `InvalidConfigurationException`

The later one will be used in places like: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/Symfony/CS/Console/ConfigurationResolver.php#L438

**Comment by keradus** [comment]
ping @SpacePossum ;)

**Comment by keradus** [comment]
@SpacePossum Please, could you focus on that?

**Comment by SpacePossum** [comment]
yes, on it :)

**Comment by keradus** [comment]
Great, thanks!

File: Symfony/CS/ConfigurationResolver.php [hunk 1]
```diff
@@ -187,8 +221,6 @@ protected function parseLevel()
                 return $this->config->getLevel();
             }
         }
-
-        return;
```
Comment (SpacePossum) [inline Symfony/CS/ConfigurationResolver.php]: fixer for the is PR'ed ;)

**Comment by SpacePossum** [comment]
I left the output as is (i.e. leave it to the symfony/application) for now to prevent make an epic big PR out of this one.
I made the two exception classes and which set the exit codes, let me know :)

File: README.rst [hunk 2]
```diff
@@ -778,6 +778,15 @@ speed up further runs.
         ->setUsingCache(true)
     ;
 
+Exit codes
+-------
```
Comment (keradus) [inline README.rst]:
```diff
 Exit codes
--------
+----------
```

Comment (SpacePossum) [inline README.rst]: :+1: done

**Comment by keradus** [comment]
Please fix the build

**Comment by SpacePossum** [comment]
tests are happy, fabbot.io is off (as we know :) )

File: Symfony/CS/ConfigurationResolver.php [hunk 3]
```diff
@@ -158,6 +176,25 @@ protected function resolveByNames()
         }
     }
 
+    protected function resolveFormat()
+    {
+        if (array_key_exists('format', $this->options)) {
+            $format = $this->options['format'];
+        } elseif (method_exists($this->config, 'getFormat')) {
+            $format = $this->options->getFormat();
```
Comment (SpacePossum) [inline Symfony/CS/ConfigurationResolver.php]: oops, this is wrong...

File: Symfony/CS/ConfigurationException/InvalidConfigurationException.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Symfony\CS\ConfigurationException;
+
+/**
+ * Exceptions of this type are thrown by the Fixer on misconfiguration.
```
Comment (keradus) [inline Symfony/CS/ConfigurationException/InvalidConfigurationException.php]: I dont think this is true

File: Symfony/CS/Console/Command/FixCommand.php [hunk 5]
```diff
@@ -428,7 +439,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
                 }
 
                 $fixEvent = $this->stopwatch->getEvent('fixFiles');
-                $output->writeln(sprintf('%s all files in %.3f seconds, %.3f MB memory used', $input->getOption('dry-run') ? 'Checked' : 'Fixed', $fixEvent->getDuration() / 1000, $fixEvent->getMemory() / 1024 / 1024));
+                $output->writeln(sprintf('%s all files in %.3f seconds, %.3f MB memory used', $input->getOption('dry-run') && !$stdin ? 'Checked' : 'Fixed', $fixEvent->getDuration() / 1000, $fixEvent->getMemory() / 1024 / 1024));
```
Comment (keradus) [inline Symfony/CS/Console/Command/FixCommand.php]: ?

File: Symfony/CS/ConfigurationException/InvalidFixerConfigurationException.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+ * @internal
+ */
+class InvalidFixerConfigurationException extends InvalidConfigurationException
+{
+    /**
+     * @var string
+     */
+    private $fixerName;
+
+    public function __construct($fixerName, $message)
```
Comment (keradus) [inline Symfony/CS/ConfigurationException/InvalidFixerConfigurationException.php]: typehinting for constructor is missing

File: Symfony/CS/ConfigurationException/InvalidFixerConfigurationException.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        parent::__construct($message, 8);
+        $this->fixerName = $fixerName;
+    }
+
+    /**
+     * @return string
+     */
+    public function getFixerName()
```
Comment (keradus) [inline Symfony/CS/ConfigurationException/InvalidFixerConfigurationException.php]: this is not used...

File: README.rst [hunk 8]
```diff
@@ -778,6 +780,13 @@ speed up further runs.
         ->setUsingCache(true)
     ;
 
+Exit codes
+----------
+* 0 OK
+* 1 No changes made
+* 4 Configuration error of the application
+* 8 Configuration error of a Fixer
```
Comment (keradus) [inline README.rst]: On 2.x line we already have:
```php
    const EXIT_STATUS_FLAG_HAS_INVALID_FILES = 4;
    const EXIT_STATUS_FLAG_HAS_CHANGED_FILES = 8;
```

For that reason, please, could you use 16 and 32 ?

File: Symfony/CS/Console/Command/FixCommand.php (hunk 5 above)
Comment (keradus) [inline Symfony/CS/Console/Command/FixCommand.php]: I just didn't know what was that for

File: Symfony/CS/ConfigurationException/InvalidFixerConfigurationException.php (hunk 7 above)
Comment (keradus) [inline Symfony/CS/ConfigurationException/InvalidFixerConfigurationException.php]: ping @SpacePossum

File: Symfony/CS/Console/Command/FixCommand.php (hunk 5 above)
Comment (SpacePossum) [inline Symfony/CS/Console/Command/FixCommand.php]: it is out of scope, I'll get back on this in a diff PR some other time ;)

**Comment by SpacePossum** [comment]
This is still missing some utest so consider it WIP.

Ping @stof for a second opion (and ping @keradus for a first :) )  about the `stdout` vs. `stderr` changes

File: Symfony/CS/Console/Command/FixCommand.php (hunk 5 above)
Comment (keradus) [inline Symfony/CS/Console/Command/FixCommand.php]: ok

**Comment by keradus** [comment]
I would handle the stdout vs stderr discussion separate. We know we need to do this, we know we want to do this, no need to do this in this PR.

Please finish it, I would really like to have it on 2.x line ;)

**Comment by SpacePossum** [comment]
check, reverted and cleanupped

File: Symfony/CS/ConfigurationException/InvalidFixerConfigurationException.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+class InvalidFixerConfigurationException extends InvalidConfigurationException
+{
+    /**
+     * @param string $fixerName
+     * @param string $message
+     */
+    public function __construct($fixerName, $message)
+    {
+        parent::__construct(sprintf('[%s] %s', $fixerName, $message), FixCommand::EXIT_STATUS_FLAG_HAS_INVALID_FIXER_CONFIG);
+        $this->fixerName = $fixerName;
```
Comment (keradus) [inline Symfony/CS/ConfigurationException/InvalidFixerConfigurationException.php]: ?

File: Symfony/CS/ConfigurationResolver.php [hunk 10]
```diff
@@ -158,6 +176,26 @@ protected function resolveByNames()
...
     }
 
+    protected function resolveFormat()
+    {
+        if (array_key_exists('format', $this->options)) {
+            $format = $this->options['format'];
+        } elseif (method_exists($this->config, 'getFormat')) {
+            $format = $this->config->getFormat();
+        } else {
+            $format = 'txt'; // fast return default
```
Comment (keradus) [inline Symfony/CS/ConfigurationResolver.php]: it is not a return ;)

File: Symfony/CS/ConfigurationResolver.php [hunk 11]
```diff
@@ -158,6 +176,26 @@ protected function resolveByNames()
...
+        }
+
+        static $formats = array('txt', 'xml', 'json');
+        if (!in_array($format, $formats, true)) {
+            throw new InvalidConfigurationException(sprintf('The format "%s" is not defined, supported are %s.', $format, implode(', ', $formats)));
+        }
+
+        $this->format = $format;
+
+        return $format;
```
Comment (keradus) [inline Symfony/CS/ConfigurationResolver.php]: no other `resolve*` function does it

Comment (SpacePossum) [inline Symfony/CS/ConfigurationResolver.php]: well those should be fixed as well than ;)

Comment (SpacePossum) [inline Symfony/CS/ConfigurationResolver.php]: reverted

File: Symfony/CS/ConfigurationResolver.php (hunk 10 above)
Comment (SpacePossum) [inline Symfony/CS/ConfigurationResolver.php]: reverted :)

File: Symfony/CS/ConfigurationException/InvalidFixerConfigurationException.php (hunk 9 above)
Comment (SpacePossum) [inline Symfony/CS/ConfigurationException/InvalidFixerConfigurationException.php]: reverted :)

File: Symfony/CS/Console/Command/FixCommand.php [hunk 12]
```diff
@@ -394,13 +413,22 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+                $fixerDetailLine = false;
+                if (OutputInterface::VERBOSITY_VERBOSE <= $verbosity) {
+                    $fixerDetailLine = $output->isDecorated() ? ' (<comment>%s</comment>)' : ' %s';
+                }
+
                 foreach ($changed as $file => $fixResult) {
-                    $output->write(sprintf('%4d) %s', $i++, $file));
+                    if ($fixerDetailLine) {
+                        $output->write(sprintf('%4d) %s', $i++, $file));
+                        if ($fixerDetailLine) {
```
Comment (keradus) [inline Symfony/CS/Console/Command/FixCommand.php]:
```php
if ($fixerDetailLine) {
    if ($fixerDetailLine) {
    }
}
```

Comment (SpacePossum) [inline Symfony/CS/Console/Command/FixCommand.php]: :+1: fixed

**Comment by keradus** [comment]
:+1:

**Comment by keradus** [comment]
Thank you @SpacePossum.

**Comment by SpacePossum** [comment]
nice, great to see this in :) 
hope you squashed the `lala` out of the messages :P

File: Symfony/CS/ConfigurationException/InvalidFixerConfigurationException.php [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+ */
+class InvalidFixerConfigurationException extends InvalidConfigurationException
+{
+    /**
+     * @param string $fixerName
+     * @param string $message
+     */
+    public function __construct($fixerName, $message)
+    {
+        parent::__construct(sprintf('[%s] %s', $fixerName, $message), FixCommand::EXIT_STATUS_FLAG_HAS_INVALID_FIXER_CONFIG);
```
Comment (gharlan) [inline Symfony/CS/ConfigurationException/InvalidFixerConfigurationException.php]: @SpacePossum here you pass two arguments, but parent constructor (`InvalidConfigurationException`) has only one parameter..

**Pull request closed without merging**